sonar-plugin-api/src/main/java/org/sonar/api/config/Settings.java

```diff
@@ -83,89 +87,130 @@
  * </pre>
  * </li>
  * </ul>
- * 
+ *
+ * History - this class is abstract since 6.1.
  * @since 2.12
  */
 @ScannerSide
 @ServerSide
 @ComputeEngineSide
-public class Settings {
+public abstract class Settings {
 
-  protected Map<String, String> properties;
-  protected PropertyDefinitions definitions;
-  private Encryption encryption;
+  private final PropertyDefinitions definitions;
+  private final Encryption encryption;
 
-  public Settings() {
-    this(new PropertyDefinitions());
+  protected Settings(PropertyDefinitions definitions, Encryption encryption) {
+    this.definitions = requireNonNull(definitions);
+    this.encryption = requireNonNull(encryption);
   }
 
-  public Settings(PropertyDefinitions definitions) {
-    this.properties = Maps.newHashMap();
-    this.definitions = definitions;
-    this.encryption = new Encryption(null);
+  protected abstract Optional<String> get(String key);
+
+  protected abstract void set(String key, String value);
+
+  protected abstract void remove(String key);
+
+  /**
+   * Immutable map of the properties that have non-default values.
+   * The default values defined by {@link PropertyDefinitions} are ignored,
+   * so the returned values are not the effective values. Basically only
+   * the non-empty results of {@link #getRawString(String)} are returned.
+   * <p>
+   * Values are not decrypted if they are encrypted with a secret key.
+   * </p>
+   */
+  public abstract Map<String, String> getProperties();
+
+  // FIXME scope to be replaced by "protected" as soon as not used by JRubyFacade
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1134)